zcash_transparent/src/address.rs

```diff
@@ -1,387 +1,57 @@
-            0xad => Some(OpCode::CheckSigVerify),
-            0xae => Some(OpCode::CheckMultisig),
-            0xaf => Some(OpCode::CheckMultisigVerify),
-            0xb0 => Some(OpCode::Nop1),
-            0xb1 => Some(OpCode::CheckLockTimeVerify),
-            0xb2 => Some(OpCode::Nop3),
-            0xb3 => Some(OpCode::Nop4),
-            0xb4 => Some(OpCode::Nop5),
-            0xb5 => Some(OpCode::Nop6),
-            0xb6 => Some(OpCode::Nop7),
-            0xb7 => Some(OpCode::Nop8),
-            0xb8 => Some(OpCode::Nop9),
-            0xb9 => Some(OpCode::Nop10),
-            0xff => Some(OpCode::Invalid),
-            _ => None,
-        }
-    }
+pub fn read_script_sig<R: Read>(mut reader: R) -> io::Result<script::Component<Opcode>> {
+    let script = Vector::read(&mut reader, |r| {
+        let mut bytes = [0; 1];
+        r.read_exact(&mut bytes).map(|_| bytes[0])
+    })?;
+    // TODO: This parser is too strict.
+    script::Component::parse(&script::Code(&script))
+        .map_err(|_| io::Error::new(io::ErrorKind::InvalidData, "invalid script_sig"))
 }
 
-/// A serialized script, used inside transparent inputs and outputs of a transaction.
-#[derive(Clone, Default, PartialEq, Eq)]
-pub struct Script(pub Vec<u8>);
-
-impl fmt::Debug for Script {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        struct ScriptPrinter<'s>(&'s [u8]);
-        impl fmt::Debug for ScriptPrinter<'_> {
-            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-                let mut l = f.debug_list();
-                let mut unknown: Option<String> = None;
-                for b in self.0 {
-                    if let Some(opcode) = OpCode::parse(*b) {
-                        if let Some(s) = unknown.take() {
-                            l.entry(&s);
-                        }
-                        l.entry(&opcode);
-                    } else {
-                        let encoded = format!("{b:02x}");
-                        if let Some(s) = &mut unknown {
-                            s.push_str(&encoded);
-                        } else {
-                            unknown = Some(encoded);
-                        }
-                    }
-                }
-                l.finish()
-            }
-        }
-
-        if f.alternate() {
-            f.debug_tuple("Script")
-                .field(&ScriptPrinter(&self.0))
-                .finish()
-        } else {
-            f.debug_tuple("Script")
-                .field(&hex::encode(&self.0))
-                .finish()
-        }
-    }
+pub fn read_script_pubkey<R: Read>(mut reader: R) -> io::Result<script::PubKey> {
+    let script = Vector::read(&mut reader, |r| {
+        let mut bytes = [0; 1];
+        r.read_exact(&mut bytes).map(|_| bytes[0])
+    })?;
+    // TODO: This parser is too strict.
+    script::PubKey::parse(&script::Code(&script))
+        .map_err(|_| io::Error::new(io::ErrorKind::InvalidData, "invalid script_pubkey"))
```

Comment:
The first commit of this PR is failing tests here, because the generated test vectors from https://github.com/zcash/zcash-test-vectors include "bad" opcodes (which are valid to be encoded, but only bad if executed). This matches the comment I added (and the documentation of the method itself) that `script::PubKey::parse` is too strict and can't be used here.